src/cargo/ops/cargo_add/mod.rs

```diff
@@ -313,8 +313,11 @@ fn resolve_dependency(
             selected
         } else {
             let source = crate::sources::PathSource::new(&path, src.source_id()?, config);
-            let packages = source.read_packages()?;
-            let package = infer_package(packages, &src, false)?;
+            let mut packages = source.read_packages()?;
+            if packages.len() == 0 {
+                anyhow::bail!("no packages found at `{src}`");
+            }
+            let package = packages.pop().expect("at least one package");
```

Comment:
Seems like this could just be `let package = packages.pop().ok_or_else(|| anyhow::bail!(...))`, no `expect` needed